File: kant_dao/src/kant_dao/mongo_dao/mongo_pddl_proposition_dao.cpp


#include <cstdint>
#include <string>
#include <vector>

#include "kant_dto/pddl_type_dto.hpp"

#include "kant_dao/mongo_dao/mongo_pddl_proposition_dao.hpp"

using namespace kant::dao::mongo_dao;

MongoPddlPropositionDao::MongoPddlPropositionDao(std::string mongo_uri,
                                                 bool init_instance)
    : MongoDao(mongo_uri, "pddl_proposition", init_instance) {

  this->mongo_pddl_object_dao = new MongoPddlObjectDao(this);
  this->mongo_pddl_predicate_dao = new MongoPddlPredicateDao(this);
}

MongoPddlPropositionDao::MongoPddlPropositionDao(
    MongoPddlObjectDao *mongo_pddl_object_dao)
    : MongoDao(mongo_pddl_object_dao->get_uri(), "pddl_proposition", false) {

  // upstream propagation
  this->mongo_pddl_object_dao = mongo_pddl_object_dao;
  this->mongo_pddl_predicate_dao = new MongoPddlPredicateDao(this);
}

MongoPddlPropositionDao::MongoPddlPropositionDao(
    MongoPddlPredicateDao *mongo_pddl_predicate_dao)
    : MongoDao(mongo_pddl_predicate_dao->get_uri(), "pddl_proposition", false) {

  // upstream propagation
  this->mongo_pddl_object_dao = new MongoPddlObjectDao(this);
  this->mongo_pddl_predicate_dao = mongo_pddl_predicate_dao;
}

MongoPddlPropositionDao::~MongoPddlPropositionDao() {
  delete this->mongo_pddl_object_dao;
  delete this->mongo_pddl_predicate_dao;
}

bsoncxx::document::value
MongoPddlPropositionDao::dto_to_mongo(kant::dto::Dto *dto) {

  kant::dto::PddlPropositionDto *pddl_proposition_dto =
      static_cast<kant::dto::PddlPropositionDto *>(dto);

  bsoncxx::builder::basic::document basic_builder{};
  basic_builder.append(bsoncxx::builder::basic::kvp(
      "predicate", pddl_proposition_dto->get_predicate()->get_name()));

  bsoncxx::builder::basic::array array_builder{};

  for (const auto &pddl_object_dto : pddl_proposition_dto->get_objects()) {
    array_builder.append(pddl_object_dto->get_name());
  }

  basic_builder.append(bsoncxx::builder::basic::kvp("objects", array_builder));

  basic_builder.append(bsoncxx::builder::basic::kvp(
      "is_goal", pddl_proposition_dto->get_is_goal()));

  return basic_builder.extract();
}

std::shared_ptr<kant::dto::PddlPropositionDto>
MongoPddlPropositionDao::mongo_to_dto(bsoncxx::document::view doc_value) {

  // PREDICATE
  std::string predicate_name =
      std::string(doc_value["predicate"].get_string().value);

  auto pddl_predicate_dto_list =
      this->mongo_pddl_predicate_dao->get(predicate_name);

  // OBJECTS
  std::vector<std::shared_ptr<kant::dto::PddlObjectDto>> pddl_object_dto_list;
  std::string object_name;

  for (const auto &mongo_pddl_object : doc_value["objects"].get_array().value) {

    object_name = std::string(mongo_pddl_object.get_string().value);

    auto pddl_object_dto = this->mongo_pddl_object_dao->get(object_name);

    pddl_object_dto_list.push_back(pddl_object_dto);
  }

  // PROPOSITIONS
  std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto =
      std::make_shared<kant::dto::PddlPropositionDto>(
          kant::dto::PddlPropositionDto(pddl_predicate_dto_list,
                                        pddl_object_dto_list));

  pddl_proposition_dto->set_is_goal(doc_value["is_goal"].get_bool());

  return pddl_proposition_dto;
}

bool MongoPddlPropositionDao::validate_dto(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {

  auto pddl_predicate_dto = pddl_proposition_dto->get_predicate();

  if (pddl_proposition_dto->get_objects().size() !=
      pddl_predicate_dto->get_types().size()) {
    return false;
  }

  for (long unsigned int i = 0; i < pddl_proposition_dto->get_objects().size();
       i++) {

    if (!pddl_proposition_dto->get_objects().at(i)->get_type()->equals(
            pddl_predicate_dto->get_types().at(i))) {
      return false;
    }
  }

  return true;
}

bool MongoPddlPropositionDao::exist_in_mongo(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {

  for (const auto &aux_pddl_proposition_dto : this->get_by_predicate(
           pddl_proposition_dto->get_predicate()->get_name())) {

    if (pddl_proposition_dto->equals(aux_pddl_proposition_dto)) {
      return true;
    }
  }

  return false;
}

std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
MongoPddlPropositionDao::get_by_predicate(std::string predicate_name) {

  mongocxx::cursor cursor =
      this->get_collection().find(bsoncxx::builder::basic::make_document(
          bsoncxx::builder::basic::kvp("predicate", predicate_name)));

  std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
      pddl_proposition_dto_list;

  for (auto doc : cursor) {
    auto pddl_proposition_dto = this->mongo_to_dto(doc);

    if (this->validate_dto(pddl_proposition_dto)) {
      pddl_proposition_dto_list.push_back(pddl_proposition_dto);
    }
  }

  return pddl_proposition_dto_list;
}

std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
MongoPddlPropositionDao::get_goals() {

  mongocxx::cursor cursor =
      this->get_collection().find(bsoncxx::builder::basic::make_document(
          bsoncxx::builder::basic::kvp("is_goal", true)));

  std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
      pddl_proposition_dto_list;

  for (auto doc : cursor) {
    auto pddl_proposition_dto = this->mongo_to_dto(doc);

    if (this->validate_dto(pddl_proposition_dto)) {
      pddl_proposition_dto_list.push_back(pddl_proposition_dto);
    }
  }

  return pddl_proposition_dto_list;
}

std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
MongoPddlPropositionDao::get_no_goals() {

  mongocxx::cursor cursor =
      this->get_collection().find(bsoncxx::builder::basic::make_document(
          bsoncxx::builder::basic::kvp("is_goal", false)));

  std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
      pddl_proposition_dto_list;

  for (auto doc : cursor) {
    auto pddl_proposition_dto = this->mongo_to_dto(doc);

    if (this->validate_dto(pddl_proposition_dto)) {
      pddl_proposition_dto_list.push_back(pddl_proposition_dto);
    }
  }
  return pddl_proposition_dto_list;
}

std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
MongoPddlPropositionDao::get_all() {

  mongocxx::cursor cursor = this->get_collection().find({});

  std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
      pddl_proposition_dto_list;

  for (auto doc : cursor) {
    pddl_proposition_dto_list.push_back(this->mongo_to_dto(doc));
  }

  return pddl_proposition_dto_list;
}

bool MongoPddlPropositionDao::save(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {

  if (!this->validate_dto(pddl_proposition_dto)) {
    return false;
  }

  if (this->exist_in_mongo(pddl_proposition_dto)) {
    return false;
  }

  bool succ = this->propagate_saving(pddl_proposition_dto);

  if (!succ) {
    return false;
  }

  return this->mongo_insert(pddl_proposition_dto.get());
}

bool MongoPddlPropositionDao::update(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {

  if (!this->validate_dto(pddl_proposition_dto)) {
    return false;
  }

  if (!this->exist_in_mongo(pddl_proposition_dto)) {
    return false;
  }

  bool succ = this->propagate_saving(pddl_proposition_dto);

  if (!succ) {
    return false;
  }

  bsoncxx::builder::basic::array array_builder{};
  for (const auto &pddl_object_dto : pddl_proposition_dto->get_objects()) {
    array_builder.append(pddl_object_dto->get_name());
  }

  this->get_collection().update_one(
      // filter
      bsoncxx::builder::basic::make_document(
          bsoncxx::builder::basic::kvp(
              "predicate", pddl_proposition_dto->get_predicate()->get_name()),
          bsoncxx::builder::basic::kvp("objects", array_builder)),
      // new data
      bsoncxx::builder::basic::make_document(bsoncxx::builder::basic::kvp(
          "$set", this->dto_to_mongo(pddl_proposition_dto.get()))));

  return true;
}

bool MongoPddlPropositionDao::propagate_saving(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {

  // PROPAGATING SAVING
  bool succ;

  // PREDICATE
  succ = this->mongo_pddl_predicate_dao->save_update(
      pddl_proposition_dto->get_predicate());

  if (!succ) {
    return false;
  }

  // OBJECTS
  for (const auto &pddl_type_dto : pddl_proposition_dto->get_objects()) {
    succ = this->mongo_pddl_object_dao->save_update(pddl_type_dto);

    if (!succ) {
      return false;
    }
  }

  return true;
}

bool MongoPddlPropositionDao::save_update(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {

  if (!this->exist_in_mongo(pddl_proposition_dto)) {
    return this->save(pddl_proposition_dto);
  } else {
    return this->update(pddl_proposition_dto);
  }
}

bool MongoPddlPropositionDao::delete_one(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {

  if (!this->exist_in_mongo(pddl_proposition_dto)) {
    return false;
  }

  return this->mongo_delete(pddl_proposition_dto.get());
}

bool MongoPddlPropositionDao::delete_all() {
  this->get_collection().delete_many({});
  return true;
}

File: kant_dao/kant_dao/mongo_dao/mongo_pddl_proposition_dao.py
# Copyright (C) 2023 Miguel Ángel González Santamarta

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


"""Mongo Pddl Proposition Dao"""

from typing import List

from kant_dao.dao_interface import PddlPropositionDao
from kant_dao.mongo_dao import MongoDao, MongoPddlPredicateDao, MongoPddlObjectDao

from kant_dao.mongo_dao.mongo_models import PddlPropositionModel, PddlPredicateModel

from kant_dto import PddlPropositionDto


class MongoPddlPropositionDao(PddlPropositionDao, MongoDao):
    """Mongo Pddl Proposition Dao Class"""

    def __init__(self, uri: str = None, connect: bool = True):

        PddlPropositionDao.__init__(self)
        MongoDao.__init__(self, uri)

        if connect:
            self.connect()

        self._me_pddl_object_dao = MongoPddlObjectDao(uri, connect=False)
        self._me_pddl_predicate_dao = MongoPddlPredicateDao(uri, connect=False)

    def _check_pddl_proposition_model(
        self, pddl_proposition_model: PddlPropositionModel
    ) -> bool:
        """check if the types of the objects of a pddl proposition model are
            the same as the types of its predicate

        Args:
            pddl_proposition_model (PddlPropositionModel): Mongoengine pddl proposition document

        Returns:
            bool: poposition is correct?
        """

        # check if proposition is correct
        if len(pddl_proposition_model.pddl_objects) != len(
            pddl_proposition_model.predicate.types
        ):
            return False

        pddl_object_models = pddl_proposition_model.pddl_objects
        pddl_type_models = pddl_proposition_model.predicate.types

        for pddl_object_model, pddl_type_model in zip(
            pddl_object_models, pddl_type_models
        ):

            # check if proposition is correct
            if pddl_object_model.type.name != pddl_type_model.name:
                return False

        return True

    def _model_to_dto(
        self, pddl_proposition_model: PddlPropositionModel
    ) -> PddlPropositionDto:
        """convert a Mongoengine pddl type document into a PddlPropositionDto

        Args:
            pddl_proposition_model (PddlPropositionModel): Mongoengine pddl proposition document

        Returns:
            PddlPropositionDto: PddlPropositionDto
        """

        pddl_object_list = []

        pddl_predicate_dto = self._me_pddl_predicate_dao._model_to_dto(
            pddl_proposition_model.predicate
        )

        for pddl_object_model in pddl_proposition_model.pddl_objects:

            pddl_object_dto = self._me_pddl_object_dao._model_to_dto(pddl_object_model)

            # check if object exist
            if not pddl_object_dto:
                return None

            pddl_object_list.append(pddl_object_dto)

        pddl_proposition_dto = PddlPropositionDto(pddl_predicate_dto, pddl_object_list)

        pddl_proposition_dto.set_is_goal(pddl_proposition_model.is_goal)

        return pddl_proposition_dto

    def _check_pddl_proposition_dto(
        self, pddl_proposition_dto: PddlPropositionDto
    ) -> bool:
        """check if the types of the objects of a pddl proposition dto are
            the same as the types of its predicate

        Args:
            pddl_proposition_dto (PddlPropositionDto): pddl proposition dto

        Returns:
            bool: poposition is correct?
        """

        # check if proposition is correct
        if len(pddl_proposition_dto.get_objects()) != len(
            pddl_proposition_dto.get_predicate().get_types()
        ):
            return False

        pddl_object_dtos = pddl_proposition_dto.get_objects()
        pddl_type_dtos = pddl_proposition_dto.get_predicate().get_types()

        for pddl_object_dto, pddl_type_dto in zip(pddl_object_dtos, pddl_type_dtos):

            # check if proposition is correct
            if pddl_object_dto.get_type() != pddl_type_dto:
                return False

        return True

    def _dto_to_model(
        self, pddl_proposition_dto: PddlPropositionDto
    ) -> PddlPropositionModel:
        """convert a PddlPropositionDto into a Mongoengine pddl propostion document

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto

        Returns:
            Document: Mongoengine pddl propostion document
        """

        pddl_proposition_model = PddlPropositionModel()

        # predicate model
        pddl_predicate_model = self._me_pddl_predicate_dao._dto_to_model(
            pddl_proposition_dto.get_predicate()
        )

        pddl_proposition_model.predicate = pddl_predicate_model

        # is goal
        pddl_proposition_model.is_goal = pddl_proposition_dto.get_is_goal()

        # objects models
        for pddl_object_dto in pddl_proposition_dto.get_objects():

            pddl_object_model = self._me_pddl_object_dao._dto_to_model(pddl_object_dto)

            pddl_proposition_model.pddl_objects.append(pddl_object_model)

        return pddl_proposition_model

    def _exist_in_mongo(self, pddl_proposition_dto: PddlPropositionDto) -> bool:
        """check if PddlPropositionDto exists

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto

        Returns:
            bool: PddlPropositionDto exists?
        """

        if self._get_model(pddl_proposition_dto):
            return True
        return False

    def _get_model(
        self, pddl_proposition_dto: PddlPropositionDto
    ) -> PddlPropositionModel:
        """get the Mongoengine pddl proposition document corresponding to a give PddlPropositionDto

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto

        Returns:
            Document: Mongoengine pddl proposition document
        """

        pddl_objects_dto_list = pddl_proposition_dto.get_objects()

        pddl_objects_list = []
        for pddl_object_dto in pddl_objects_dto_list:
            pddl_objects_list.append(pddl_object_dto.get_name())

        # getting proposition
        pddl_proposition_model = PddlPropositionModel.objects(
            predicate=pddl_proposition_dto.get_predicate().get_name(),
            pddl_objects=pddl_objects_list,
            is_goal=pddl_proposition_dto.get_is_goal(),
        )

        # check if proposition exist
        if not pddl_proposition_model:
            return None

        return pddl_proposition_model[0]

    def get_by_predicate(self, predicate_name: str) -> List[PddlPropositionDto]:
        """get all PddlPropositionDto with a given pddl predicate name

        Args:
            predicate_name (str): pddl predicate name

        Returns:
            List[PddlPropositionDto]: list of PddlPropositionDto
        """

        pddl_proposition_model = PddlPropositionModel.objects(predicate=predicate_name)

        pddl_proposition_dto_list = []

        for ele in pddl_proposition_model:
            if self._check_pddl_proposition_model(ele):
                pddl_proposition_dto_list.append(self._model_to_dto(ele))

        return pddl_proposition_dto_list

    def _get_all(self, is_goal: bool = None) -> List[PddlPropositionDto]:
        """get all PddlPropositionDto
            is_goal == None -> get all proposition
            is_goal == True -> gel all goals
            is_goal == False -> getl no goals

        Args:
            is_goal (bool, optional): get all, all goals, all no goals?. Defaults to None.

        Returns:
            List[PddlPropositionDto]: list of PddlPropositionDto
        """

        if is_goal is None:
            pddl_proposition_model = PddlPropositionModel.objects()
        else:
            pddl_proposition_model = PddlPropositionModel.objects(is_goal=is_goal)

        pddl_proposition_dto_list = []

        for ele in pddl_proposition_model:
            if self._check_pddl_proposition_model(ele):
                pddl_proposition_dto_list.append(self._model_to_dto(ele))

        return pddl_proposition_dto_list

    def get_goals(self) -> List[PddlPropositionDto]:
        """get all PddlPropositionDto that are goals

        Returns:
            List[PddlPropositionDto]: list of PddlPropositionDto
        """

        return self._get_all(is_goal=True)

    def get_no_goals(self) -> List[PddlPropositionDto]:
        """get all PddlPropositionDto that are not goals

        Returns:
            List[PddlPropositionDto]: list of PddlPropositionDto
        """

        return self._get_all(is_goal=False)

    def get_all(self) -> List[PddlPropositionDto]:
        """get all PddlPropositionDto

        Returns:
            List[PddlPropositionDto]: list of PddlPropositionDto
        """

        return self._get_all()

    def _save(self, pddl_proposition_dto: PddlPropositionDto) -> bool:
        """save a PddlPropositionDto
            if the PddlPropositionDto is already saved return False, else return True

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto to save

        Returns:
            bool: succeed
        """

        if self._exist_in_mongo(pddl_proposition_dto):
            return False

        if not self._check_pddl_proposition_dto(pddl_proposition_dto):
            return False

        # propagating saving
        for pddl_object_dto in pddl_proposition_dto.get_objects():
            if not self._me_pddl_object_dao.save(pddl_object_dto):
                return False

        if not self._me_pddl_predicate_dao.save(pddl_proposition_dto.get_predicate()):
            return False

        # saving
        pddl_proposition_model = self._dto_to_model(pddl_proposition_dto)

        if pddl_proposition_model:
            pddl_proposition_model.save(cascade=True)
            return True

        return False

    def _update(self, pddl_proposition_dto: PddlPropositionDto) -> bool:
        """update a PddlPropositionDto
            if the PddlPropositionDto is not saved return False, else return True

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto to update

        Returns:
            bool: succeed
        """

        if not self._check_pddl_proposition_dto(pddl_proposition_dto):
            return False

        pddl_proposition_model = self._get_model(pddl_proposition_dto)

        # check if proposition exists
        if pddl_proposition_model:
            new_pddl_proposition_model = self._dto_to_model(pddl_proposition_dto)
            pddl_proposition_model.predicate = new_pddl_proposition_model.predicate
            pddl_proposition_model.pddl_objects = new_pddl_proposition_model.pddl_objects
            pddl_proposition_model.is_goal = new_pddl_proposition_model.is_goal
            pddl_proposition_model.save()

            return True

        return False

    def save(self, pddl_proposition_dto: PddlPropositionDto) -> bool:
        """save or update a PddlPropositionDto
            if the PddlPropositionDto is not saved it will be saved, else it will be updated

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto to save or update

        Returns:
            bool: succeed
        """

        if self._exist_in_mongo(pddl_proposition_dto):
            return self._update(pddl_proposition_dto)

        return self._save(pddl_proposition_dto)

    def delete(self, pddl_proposition_dto: PddlPropositionDto) -> bool:
        """delete a PddlPropositionDto
            if the PddlPropositionDto is not saved return False, else return True

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto to delete

        Returns:
            bool: succeed
        """

        pddl_proposition_model = self._get_model(pddl_proposition_dto)

        # check if proposition exists
        if pddl_proposition_model:
            pddl_proposition_model.delete()
            return True

        return False

    def delete_all(self) -> bool:
        """delete all pddl propositions

        Returns:
            bool: succeed
        """

        pddl_proposition_dto_list = self.get_all()

        for ele in pddl_proposition_dto_list:
            self.delete(ele)

        return True
